File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-ird-5-4/Discounting.cpp
// Discounting.cpp 
#include "Discounting.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Discounting::Discounting(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //discountingTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* discountingTypeNode = xmlNode->FirstChildElement("discountingType");

   if(discountingTypeNode){discountingTypeIsNull_ = false;}
   else{discountingTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- discountingTypeNode , address : " << discountingTypeNode << std::endl;
   #endif
   if(discountingTypeNode)
   {
      if(discountingTypeNode->Attribute("href") || discountingTypeNode->Attribute("id"))
      {
          if(discountingTypeNode->Attribute("id"))
          {
             discountingTypeIDRef_ = discountingTypeNode->Attribute("id");
             discountingType_ = boost::shared_ptr<DiscountingTypeEnum>(new DiscountingTypeEnum(discountingTypeNode));
             discountingType_->setID(discountingTypeIDRef_);
             IDManager::instance().setID(discountingTypeIDRef_,discountingType_);
          }
          else if(discountingTypeNode->Attribute("href")) { discountingTypeIDRef_ = discountingTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { discountingType_ = boost::shared_ptr<DiscountingTypeEnum>(new DiscountingTypeEnum(discountingTypeNode));}
   }

   //discountRateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* discountRateNode = xmlNode->FirstChildElement("discountRate");

   if(discountRateNode){discountRateIsNull_ = false;}
   else{discountRateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- discountRateNode , address : " << discountRateNode << std::endl;
   #endif
   if(discountRateNode)
   {
      if(discountRateNode->Attribute("href") || discountRateNode->Attribute("id"))
      {
          if(discountRateNode->Attribute("id"))
          {
             discountRateIDRef_ = discountRateNode->Attribute("id");
             discountRate_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(discountRateNode));
             discountRate_->setID(discountRateIDRef_);
             IDManager::instance().setID(discountRateIDRef_,discountRate_);
          }
          else if(discountRateNode->Attribute("href")) { discountRateIDRef_ = discountRateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { discountRate_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(discountRateNode));}
   }

   //discountRateDayCountFractionNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* discountRateDayCountFractionNode = xmlNode->FirstChildElement("discountRateDayCountFraction");

   if(discountRateDayCountFractionNode){discountRateDayCountFractionIsNull_ = false;}
   else{discountRateDayCountFractionIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- discountRateDayCountFractionNode , address : " << discountRateDayCountFractionNode << std::endl;
   #endif
   if(discountRateDayCountFractionNode)
   {
      if(discountRateDayCountFractionNode->Attribute("href") || discountRateDayCountFractionNode->Attribute("id"))
      {
          if(discountRateDayCountFractionNode->Attribute("id"))
          {
             discountRateDayCountFractionIDRef_ = discountRateDayCountFractionNode->Attribute("id");
             discountRateDayCountFraction_ = boost::shared_ptr<DayCountFraction>(new DayCountFraction(discountRateDayCountFractionNode));
             discountRateDayCountFraction_->setID(discountRateDayCountFractionIDRef_);
             IDManager::instance().setID(discountRateDayCountFractionIDRef_,discountRateDayCountFraction_);
          }
          else if(discountRateDayCountFractionNode->Attribute("href")) { discountRateDayCountFractionIDRef_ = discountRateDayCountFractionNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { discountRateDayCountFraction_ = boost::shared_ptr<DayCountFraction>(new DayCountFraction(discountRateDayCountFractionNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<DiscountingTypeEnum> Discounting::getDiscountingType()
{
   if(!this->discountingTypeIsNull_){
        if(discountingTypeIDRef_ != ""){
             return boost::shared_static_cast<DiscountingTypeEnum>(IDManager::instance().getID(discountingTypeIDRef_));
        }else{
             return this->discountingType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<DiscountingTypeEnum>();
   }
}
boost::shared_ptr<XsdTypeDecimal> Discounting::getDiscountRate()
{
   if(!this->discountRateIsNull_){
        if(discountRateIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(discountRateIDRef_));
        }else{
             return this->discountRate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
boost::shared_ptr<DayCountFraction> Discounting::getDiscountRateDayCountFraction()
{
   if(!this->discountRateDayCountFractionIsNull_){
        if(discountRateDayCountFractionIDRef_ != ""){
             return boost::shared_static_cast<DayCountFraction>(IDManager::instance().getID(discountRateDayCountFractionIDRef_));
        }else{
             return this->discountRateDayCountFraction_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<DayCountFraction>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-ird-5-4/Discounting.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Discounting : ISerialized
    {
        public Discounting(XmlNode xmlNode)
        {
            XmlNode discountingTypeNode = xmlNode.SelectSingleNode("discountingType");
            
            if (discountingTypeNode != null)
            {
                if (discountingTypeNode.Attributes["href"] != null || discountingTypeNode.Attributes["id"] != null) 
                {
                    if (discountingTypeNode.Attributes["id"] != null) 
                    {
                        discountingTypeIDRef_ = discountingTypeNode.Attributes["id"].Value;
                        DiscountingTypeEnum ob = new DiscountingTypeEnum(discountingTypeNode);
                        IDManager.SetID(discountingTypeIDRef_, ob);
                    }
                    else if (discountingTypeNode.Attributes["href"] != null)
                    {
                        discountingTypeIDRef_ = discountingTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        discountingType_ = new DiscountingTypeEnum(discountingTypeNode);
                    }
                }
                else
                {
                    discountingType_ = new DiscountingTypeEnum(discountingTypeNode);
                }
            }
            
        
            XmlNode discountRateNode = xmlNode.SelectSingleNode("discountRate");
            
            if (discountRateNode != null)
            {
                if (discountRateNode.Attributes["href"] != null || discountRateNode.Attributes["id"] != null) 
                {
                    if (discountRateNode.Attributes["id"] != null) 
                    {
                        discountRateIDRef_ = discountRateNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(discountRateNode);
                        IDManager.SetID(discountRateIDRef_, ob);
                    }
                    else if (discountRateNode.Attributes["href"] != null)
                    {
                        discountRateIDRef_ = discountRateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        discountRate_ = new XsdTypeDecimal(discountRateNode);
                    }
                }
                else
                {
                    discountRate_ = new XsdTypeDecimal(discountRateNode);
                }
            }
            
        
            XmlNode discountRateDayCountFractionNode = xmlNode.SelectSingleNode("discountRateDayCountFraction");
            
            if (discountRateDayCountFractionNode != null)
            {
                if (discountRateDayCountFractionNode.Attributes["href"] != null || discountRateDayCountFractionNode.Attributes["id"] != null) 
                {
                    if (discountRateDayCountFractionNode.Attributes["id"] != null) 
                    {
                        discountRateDayCountFractionIDRef_ = discountRateDayCountFractionNode.Attributes["id"].Value;
                        DayCountFraction ob = new DayCountFraction(discountRateDayCountFractionNode);
                        IDManager.SetID(discountRateDayCountFractionIDRef_, ob);
                    }
                    else if (discountRateDayCountFractionNode.Attributes["href"] != null)
                    {
                        discountRateDayCountFractionIDRef_ = discountRateDayCountFractionNode.Attributes["href"].Value;
                    }
                    else
                    {
                        discountRateDayCountFraction_ = new DayCountFraction(discountRateDayCountFractionNode);
                    }
                }
                else
                {
                    discountRateDayCountFraction_ = new DayCountFraction(discountRateDayCountFractionNode);
                }
            }
            
        
        }
        
    
        #region DiscountingType_
        private DiscountingTypeEnum discountingType_;
        public DiscountingTypeEnum DiscountingType_
        {
            get
            {
                if (this.discountingType_ != null)
                {
                    return this.discountingType_; 
                }
                else if (this.discountingTypeIDRef_ != null)
                {
                    discountingType_ = IDManager.getID(discountingTypeIDRef_) as DiscountingTypeEnum;
                    return this.discountingType_; 
                }
                else
                {
                      return this.discountingType_; 
                }
            }
            set
            {
                if (this.discountingType_ != value)
                {
                    this.discountingType_ = value;
                }
            }
        }
        #endregion
        
        public string discountingTypeIDRef_ { get; set; }
        #region DiscountRate_
        private XsdTypeDecimal discountRate_;
        public XsdTypeDecimal DiscountRate_
        {
            get
            {
                if (this.discountRate_ != null)
                {
                    return this.discountRate_; 
                }
                else if (this.discountRateIDRef_ != null)
                {
                    discountRate_ = IDManager.getID(discountRateIDRef_) as XsdTypeDecimal;
                    return this.discountRate_; 
                }
                else
                {
                      return this.discountRate_; 
                }
            }
            set
            {
                if (this.discountRate_ != value)
                {
                    this.discountRate_ = value;
                }
            }
        }
        #endregion
        
        public string discountRateIDRef_ { get; set; }
        #region DiscountRateDayCountFraction_
        private DayCountFraction discountRateDayCountFraction_;
        public DayCountFraction DiscountRateDayCountFraction_
        {
            get
            {
                if (this.discountRateDayCountFraction_ != null)
                {
                    return this.discountRateDayCountFraction_; 
                }
                else if (this.discountRateDayCountFractionIDRef_ != null)
                {
                    discountRateDayCountFraction_ = IDManager.getID(discountRateDayCountFractionIDRef_) as DayCountFraction;
                    return this.discountRateDayCountFraction_; 
                }
                else
                {
                      return this.discountRateDayCountFraction_; 
                }
            }
            set
            {
                if (this.discountRateDayCountFraction_ != value)
                {
                    this.discountRateDayCountFraction_ = value;
                }
            }
        }
        #endregion
        
        public string discountRateDayCountFractionIDRef_ { get; set; }
        
    
        
    
    }
    
}

